astropy/io/fits/hdu/hdulist.py

```diff
@@ -179,11 +234,35 @@ def __init__(self, hdus=[], file=None):
 
         super(HDUList, self).__init__(hdus)
 
-        self.update_extend()
+        if file is None:
+            # Only do this when initializing from an existing list of HDUs
+            # When initalizing from a file, this will be handled by the
+            # append method after the first HDU is read
+            self.update_extend()
+
+    def __len__(self):
+        if not self._in_read_next_hdu:
+            while self._read_next_hdu():
+                pass
+
+        return super(HDUList, self).__len__()
+
+    def __repr__(self):
+        # In order to correctly repr an HDUList we need to load all the
+        # HDUs as well
+        while self._read_next_hdu():
+            pass
+
+        return super(HDUList, self).__repr__()
 
     def __iter__(self):
-        for idx in range(len(self)):
-            yield self[idx]
+        idx = 0
```

Comment:
needs a comment that the obvious way of implementing this does break lazy loading via len()